feat: The finite product of semi-rings (in terms of measure theory) is a semi-ring.

pfaffelh opened the pull request:
~~Move results for rings (in terms of measure theory) to a separate file; was in Semiring.lean before.~~
For `∀ i ∈ s, IsSetSemiring (C i))`, the product `s.pi '' s.pi C` is a semiring. 
Prove two auxiliary lemmas in `Data.Set.Prod` needed on the way.

pfaffelh committed:
initialized piSemiring

Commit:
at the main step

Commit:
reached interesting spot

Commit:
on step

Commit:
not done

pfaffelh committed:
Merge branch 'master' into pfaffelh_SemiringPi

Commit:
Merge branch 'master' into pfaffelh_SemiringPi

Commit:
further in the proof

pfaffelh committed:
Merge branch 'pfaffelh_SemiringPi' of https://github.com/leanprover-community/mathlib4 into pfaffelh_SemiringPi

Commit:
further in the proof2

pfaffelh committed:
Merge branch 'pfaffelh_SemiringPi' of https://github.com/leanprover-community/mathlib4 into pfaffelh_SemiringPi

Commit:
zero case done

Commit:
fintype done

Commit:
some general results

Commit:
golfed proof

Commit:
Merge branch 'master' into pfaffelh_SemiringPi

Commit:
towards PR1

Commit:
towards PR 2

Commit:
towards PR 3

Commit:
towards PR 4

Commit:
repaired imports for Ring

pfaffelh committed:
Merge branch 'pfaffelh_SemiringPi' of https://github.com/leanprover-community/mathlib4 into pfaffelh_SemiringPi

Commit:
update Mathlib.lean

Commit:
removed SetRing

Commit:
restored imports

Commit:
Merge branch 'master' into pfaffelh_SemiringPi

Paul-Lez commented on Mathlib/Data/Set/Pairwise/Lattice.lean:
```diff
@@ -55,6 +55,14 @@ theorem pairwiseDisjoint_sUnion {s : Set (Set ι)} (h : DirectedOn (· ⊆ ·) s
     (⋃₀ s).PairwiseDisjoint f ↔ ∀ ⦃a⦄, a ∈ s → Set.PairwiseDisjoint a f :=
   pairwise_sUnion h
 
+theorem pairwiseDisjoint_union_of_disjoint
+  {s t : Set (Set ι)} {f : Set ι → α} (hf : Monotone f)
+  (hs : PairwiseDisjoint s f) (ht : PairwiseDisjoint t f)
+    (hst : Disjoint (f (Set.sUnion s)) (f (sUnion t))) : PairwiseDisjoint (s ∪ t) f := by
```
```suggestion
theorem pairwiseDisjoint_union_of_disjoint
    {s t : Set (Set ι)} {f : Set ι → α} (hf : Monotone f)
    (hs : PairwiseDisjoint s f) (ht : PairwiseDisjoint t f)
    (hst : Disjoint (f (Set.sUnion s)) (f (sUnion t))) : PairwiseDisjoint (s ∪ t) f := by
```

Paul-Lez commented on Mathlib/Data/Set/Prod.lean:
```diff
@@ -730,6 +730,23 @@ theorem pi_if {p : ι → Prop} [h : DecidablePred p] (s : Set ι) (t₁ t₂ :
+  rw [← union_diff_cancel h, union_pi]
+  exact Set.inter_subset_left
+
+open scoped Classical in
+lemma union_pi_ite_of_disjoint {s t : Set ι} {x y : (i : ι) → Set (α i)} (hst : Disjoint s t) :
+((s ∪ t).pi fun i ↦ if i ∈ s then x i else y i)  = (s.pi x) ∩ (t.pi y) := by
+  have hx : ∀ i ∈ s, x i = if h : i ∈ s then x i else y i := by
+    intro i hi
+    simp only [dite_eq_ite, hi, ↓reduceIte]
+  have hy : ∀ i ∈ t, y i = if h : i ∈ s then x i else y i := by
+    intro i hi
+    have h : i ∉ s := Disjoint.not_mem_of_mem_left (id (Disjoint.symm hst)) hi
+    simp only [hi, hst, dite_eq_ite, h, ↓reduceIte]
+  rw [Set.pi_congr rfl hx, Set.pi_congr rfl hy]
+  exact union_pi
```
```suggestion
lemma union_pi_ite_of_disjoint {s t : Set ι} {x y : (i : ι) → Set (α i)} (hst : Disjoint s t)
    [DecidablePred (· ∈ s)] :
    ((s ∪ t).pi fun i ↦ if i ∈ s then x i else y i) = (s.pi x) ∩ (t.pi y) := by
  rw [union_pi, Set.pi_congr rfl (fun i hi ↦ if_pos hi), Set.pi_congr rfl (fun i hi ↦
    if_neg <| hst.symm.not_mem_of_mem_left hi)]
```

EtienneC30 commented on Mathlib/Data/Set/Prod.lean:
```diff
@@ -890,6 +907,58 @@ theorem univ_pi_ite (s : Set ι) [DecidablePred (· ∈ s)] (t : ∀ i, Set (α
+section Setdiff
+
+/-- Write that set difference `(s ∪ t).pi x \ (s ∪ t).pi y` as a union. -/
+lemma pi_setdiff_eq_union (s t : Set ι) (x y : (i : ι) → Set (α i)) :
+  (s ∪ t).pi x \ (s ∪ t).pi y = (t.pi x \ t.pi y) ∩ (s.pi x ∩ s.pi y) ∪
+    t.pi x ∩ (s.pi x \ s.pi y) := by
```
I don't think we want this lemma, it seems way too specific. You should put the proof directly in the place where you use it, and you can do a much shorter proof using [Set.union_pi](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Prod.html#Set.union_pi) (I changed a bit the right-hand side to remove a few rewrites):
```lean
lemma pi_setdiff_eq_union (s t : Set ι) (x y : (i : ι) → Set (α i)) :
  (s ∪ t).pi x \ (s ∪ t).pi y = (s.pi y)ᶜ ∩ (s.pi x ∩ t.pi x) ∪ (t.pi y)ᶜ ∩ (s.pi x ∩ t.pi x) := by
  rw [diff_eq_compl_inter, union_pi, union_pi, compl_inter, union_inter_distrib_right]
```

EtienneC30 commented:
@Paul-Lez I am guessing you pushed on the wrong branch?

Paul-Lez commented:
> @Paul-Lez I am guessing you pushed on the wrong branch?

Yes, my apologies for this! Hopefully this is now fixed!

pfaffelh committed:
Merge branch 'pfaffelh_SemiringPi' of https://github.com/leanprover-community/mathlib4 into pfaffelh_SemiringPi

pfaffelh committed:
merged master

pfaffelh committed:
shortened proof

pfaffelh committed:
shortened proof2

pfaffelh committed:
Merge remote-tracking branch 'origin/master' into pfaffelh_SemiringPi

pfaffelh committed:
shorter

pfaffelh committed:
shorter2

alreadydone committed:
chore: correct typo (#24559)

YaelDillies committed:
feat: cofinal sets (#24290)

A set `s` is dominated by a set `t` if for all `a ∈ s` there exists `b ∈ t` such that `a ≤ b`.

This naturally occuring condition is equivalent to `s ⊆ lowerClosure t`, but it is difficult to prove things like transitivity from that definition alone. Better to have API.

The import increase is immaterial.

Co-authored-by: Chris Hoskin

kim-em committed:
chore: bump toolchain to v4.20.0-rc2 (#24561)



Co-authored-by: Johan Commelin <johan@commelin.net>
Co-authored-by: Kyle Miller <kmill31415@gmail.com>
Co-authored-by: Kyle Miller <leanprover-community-mathlib4-bot@users.noreply.github.com>

Ruben-VandeVelde committed:
feat: pow_finEquivZPowers_symm_apply (#24548)

From flt-regular.

Co-authored-by: erd1 <the.erd.one@gmail.com>

robertmaxton42 committed:
feat (Small) : Shrink.rec_equivShrink (#24535)

* Adds a convenience lemma `Shrink.rec_equivShrink` that is very nearly the definitional property of `Shrink.rec`, but for whatever reason can't be automatically proven by `simp`.

urkud committed:
feat(Asymptotics): define `IsBigOTVS` (#23756)

- Turn `IsLittleOTVS` into a 1-field `structure`. We may want to replace the definition with "the ratio of `egauge`s tends to zero". In `ENNReal`, it's the right definition.
- Define `IsBigOTVS`, prove basic API lemmas.
- Prove lemmas about `IsLittleOTVS` in (indexed) products.

JovanGerb committed:
chore(Analysis/InnerProductSpace): make `𝕜` an explicit argument of `inner` (#24499)

There are two reasons why `𝕜` should be an explicit argument.
- This change makes it so that notations `⟪x, y⟫` and `⟪x, y⟫_𝕜` elaborate properly, so that this notation can be seen in the infoview.
- This change makes is so that the expected type is known whenever using `inner`. Thus, this PR also removes some now unneccessary type annotations.

The argument explicitness now matches that of `innerₛₗ` and `innerSL`.

`notation3` has been replaced by `notation`, since we don't need the more fancy `notation3` elborator.

plp127 committed:
feat(Topology/Inseparable): more mapping lemmas (#24526)

Prove `Specializes.map_of_continuousWithinAt`, `Specializes.map_of_continuousOn`, and `Inseparable.map_of_continuousWithinAt`, `Inseparable.map_of_continuousOn`.

Thmoas-Guan committed:
feat(Algebra): finiteness of associated primes (#24242)

In this PR we proved the finiteness of associated primes for finitely generated modules over notherian ring. Mainly from https://github.com/acmepjz/lean-iwasawa/blob/master/Iwasawalib/RingTheory/CharacteristicIdeal/Basic.lean authored by @acmepjz 

Co-authored-by: Jz Pan <acme_pjz@hotmail.com>



Co-authored-by: acmepjz <acme_pjz@hotmail.com>

sgouezel commented:
There is a problem with the branch (it adds 1572 lines and removes 692, and touches 69 files, which is probably not what you wanted)

urkud committed:
feat(Homotopy): +3 lemmas (#24564)

- extract `Homotopy.extend_of_mem_I` from `Homotopy.extend_apply_of_mem_I`;
- add convenience lemmas `extend_zero` and `extend_one`.

dupuisf committed:
feat(CStarAlgebra): define completely positive maps (#24359)

This PR defines completely positive maps. A linear map `φ : A₁ →ₗ[ℂ] A₂` (where `A₁` and `A₂` are C⋆-algebras) is called completely positive (CP) if applying `φ` to all entries of a k × k matrix is also positive for every `k ∈ ℕ`.

Here we provide the definition and a few basic properties: CP maps are positive, and non-unital star algebra homomorphisms are CP. A fair amount of supporting material was also added to `CStarMatrix`. The API is still very barebones and we leave e.g. composition for subsequent PRs.



Co-authored-by: Frédéric Dupuis <31101893+dupuisf@users.noreply.github.com>

urkud committed:
feat(Bernstein): generalize to normed spaces (#24566)

Generalize Bernstein approximations from `f : C(I, ℝ)` to `f : C(I, E)`,
where `E` is a real normed space.

joelriou committed:
feat(CategoryTheory): localization of linear categories (#24309)

If `L : C ⥤ D` is an additive localization functor between preadditive categories, and `C` is `R`-linear, we show that `D` can also be equipped with a `R`-linear structure such that `L` is a `R`-linear functor

YaelDillies committed:
chore: deprecate former interaction of finite intervals and succ/pred lemmas (#22327)

Moves:
* `Nat.Icc_succ_left` → `Finset.Icc_succ_left_eq_Ioc`
* `Nat.Ico_succ_right` → `Finset.Ico_succ_right_eq_Icc`
* `Nat.Ico_succ_left` → `Finset.Ico_succ_left_eq_Ioo`
* `Nat.Icc_pred_right` → `Finset.Icc_pred_right_eq_Ico`
* `Nat.Ico_succ_succ` → `Finset.Ico_succ_succ_eq_Ioc`
* `Nat.Ico_succ_right_eq_insert_Ico` → `Finset.insert_Ico_right_eq_Ico_succ_right`
* `Nat.Ico_insert_succ_left` → `Finset.insert_Ico_succ_left_eq_Ico`
* `Nat.Icc_insert_succ_left` → `Finset.insert_Icc_succ_left_eq_Icc`
* `Nat.Icc_insert_succ_right` → `Finset.insert_Icc_eq_Icc_succ_right`

Commit:
Merge branch 'master' into HEAD

pfaffelh commented:
> There is a problem with the branch (it adds 1572 lines and removes 692, and touches 69 files, which is probably not what you wanted)

Indeed, this was not what I intended, thanks for pointing this out. Now there are 5 files touched, which is what I intended.

sgouezel commented on Mathlib/Data/Set/Prod.lean:
```diff
@@ -893,6 +905,25 @@ theorem univ_pi_ite (s : Set ι) [DecidablePred (· ∈ s)] (t : ∀ i, Set (α
 
+section Setdiff
+
+/-- Write that set difference `(s ∪ t).pi x \ (s ∪ t).pi y` as a union. -/
+lemma pi_setdiff_eq_union (s t : Set ι) (x y : (i : ι) → Set (α i)) :
+  (s ∪ t).pi x \ (s ∪ t).pi y = t.pi x ∩ (s.pi x \ s.pi y) ∪ (t.pi x \ t.pi y) ∩ (s.pi x ∩ s.pi y)
+     := by
```
```suggestion
/-- Write the set difference `(s ∪ t).pi x \ (s ∪ t).pi y` as a union. -/
lemma pi_setdiff_eq_union (s t : Set ι) (x y : (i : ι) → Set (α i)) :
  (s ∪ t).pi x \ (s ∪ t).pi y = 
    t.pi x ∩ (s.pi x \ s.pi y) ∪ (t.pi x \ t.pi y) ∩ (s.pi x ∩ s.pi y) := by
```

sgouezel commented on Mathlib/Data/Set/Prod.lean:
```diff
@@ -893,6 +905,25 @@ theorem univ_pi_ite (s : Set ι) [DecidablePred (· ∈ s)] (t : ∀ i, Set (α
+  apply congrArg (Union.union (t.pi x ∩ (s.pi x \ s.pi y)))
+  aesop
+
+lemma pi_setdiff_union_disjoint (s t : Set ι) (x : (i : ι) → Set (α i)) (y : (i : ι) → Set (α i)) :
```
the lemma name should start with `disjoint` as it's the first symbol in the statement. Also, your name uses `union`, but I don't see a union anywhere.

sgouezel commented on Mathlib/Data/Set/Prod.lean:
```diff
@@ -931,6 +962,21 @@ theorem sumPiEquivProdPi_symm_preimage_univ_pi (π : ι ⊕ ι' → Type*) (t :
+variable {ι ι' : Type*} {α : ι → Type*}
+
+lemma subset_pi_image_of_subset {s : Set ι} {B C : (i : ι) → Set (Set (α i))}
+    (hBC : ∀ i ∈ s, B i ⊆ C i) : s.pi  '' s.pi B ⊆ s.pi  '' s.pi C := by
```
```suggestion
    (hBC : ∀ i ∈ s, B i ⊆ C i) : s.pi '' s.pi B ⊆ s.pi '' s.pi C := by
```

sgouezel commented on Mathlib/MeasureTheory/SetSemiring.lean:
```diff
@@ -7,8 +7,9 @@ import Mathlib.Data.Nat.Lattice
+import Mathlib.MeasureTheory.MeasurableSpace.Pi
 
-/-! # Semirings and rings of sets
+/-! # Semirings of sets
```
Why this change and the changes below? This file also defines rings of sets, if I'm not mistaken.

Commit:
Merge branch 'master' into pfaffelh_SemiringPi

pfaffelh committed:
Update Mathlib/Tactic.lean

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

pfaffelh committed:
Update Mathlib/Tactic.lean

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
resolved sth

Commit:
resolved sth2

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

Commit:
merge master

Commit:
deprecated replaced

Commit:
mk_all

Commit:
Finite2

RemyDegenne commented:
Closing: replaced by #25902
